Enable Campaigns by default (with write-access only for admins)

mrnugget opened the pull request:
This fixes https://github.com/sourcegraph/sourcegraph/issues/10713 by deprecating all the previous feature flags related to campaigns and replacing them with `campaigns.enabled: Boolean`.

By default, the value of `campaigns.enabled` is true so that campaigns are enabled for everybody, including non-site-admins.

But, in order to allow us to safely merge this while we're still working on campaigns, I left in the safety switch that prohibits users from creating changeset specs and campaign specs.

That means:

- By default site-admins can create/manage/view campaigns
- By default all users can view campaigns

When `campaigns.enabled` is set to `false`:

- site-admins cannot create/manage/view campaigns
- users cannot view campaigns

In order to then allow users to create/manage campaigns, all we need to do is to remove the two checks in `CreateChangesetSpec`/`CreateCampaignSpec` in `campaigns/resolvers/resolver.go`. The checks also have tests that can be removed.

eseliger commented on CHANGELOG.md:
```diff
@@ -15,6 +15,9 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Changed
 
+- The campaigns feature is now enabled by default for all users. It can be disabled by setting `campaigns.enabled` to `false` in the site configuration.
```
```suggestion
- The campaigns beta feature is now enabled by default for all users. It can be disabled by setting `campaigns.enabled` to `false` in the site configuration.
```
I would like to somehow make clear that it still is beta software, but not sure about the wording :)

eseliger commented on web/src/enterprise/campaigns/global/marketing/CampaignsUserMarketingPage.tsx:
```diff
@@ -1,14 +1,7 @@
-export const CampaignsUserMarketingPage: React.FunctionComponent<CampaignsUserMarketingPageProps> = ({
-    enableReadAccess,
-}) => (
+export const CampaignsUserMarketingPage: React.FunctionComponent = () => (
```
```suggestion
export const CampaignsUserMarketingPage: React.FunctionComponent<{}> = () => (
```

just to indicate it intentionally doesn't accept anything :)

eseliger commented on web/src/enterprise/campaigns/detail/CampaignDetails.tsx:
```diff
@@ -510,7 +510,7 @@ export const CampaignDetails: React.FunctionComponent<Props> = ({
                 )}
             </Form>
-            {/* Iff either campaign XOR patchset are present */}
+            {/* If either campaign XOR patchset are present */}
```
This wasn't meant to be a typo but `if and only if`, but no objection against changing it.

mrnugget replied:
TIL about [iff](https://en.wikipedia.org/wiki/If_and_only_if)!

eseliger commented on internal/conf/computed.go:
```diff
@@ -227,17 +227,11 @@ func SymbolIndexEnabled() bool {
-	// DEPRECATED property name.
-	if v := Get().AutomationReadAccessEnabled; v != nil {
+func CampaignsEnabled() bool {
+	if v := Get().SearchIndexEnabled; v != nil {
```
that seems incorrect?

mrnugget replied:
Wow 🤦‍♂️

eseliger commented:
Crap, lost my comment again. But this time I saved it:

> Should the menu item also be hidden when campaigns.enabled == false? I think so, but I want confirmation.

I think in the past we always showed it, as it shows the marketing pages and reminds people to try it out :) Do I'd keep it.

> I think we could merge this now, since all it really does is make "campaigns.readAccess" the default. But, @sqs @eseliger, since this touches the part of the frontend code that shows/hides the marketing pages, I want your input especially on this.

I think this is fine. The only step I want to take is to make the dotcom marketing page not require auth (separate PR), other than that, I don't see what would really change here :) 

BUT: It seems like "is on by default" results in: it is on on dotcom as well - right? If so, I think we need to change the check to (pseudo code) `IsCampaignsEnabled && ! IsDotcomMode`.

mrnugget commented:
> BUT: It seems like "is on by default" results in: it is on on dotcom as well - right? If so, I think we need to change the check to (pseudo code) `IsCampaignsEnabled && ! IsDotcomMode`.

Good catch! I forgot to add that! Thanks. Will add it to the resolvers check.

mrnugget commented:
Just updated it so that write AND read access to campaigns API endpoints are disabled on sourcegraph.com.

eseliger commented on CHANGELOG.md:
```diff
@@ -16,6 +16,9 @@ All notable changes to Sourcegraph are documented in this file.
 ### Changed
 
 - Some monitoring alerts now have more useful descriptions. [#11542](https://github.com/sourcegraph/sourcegraph/pull/11542)
+- The campaigns feature is now enabled by default for all users. It can be disabled by setting `campaigns.enabled` to `false` in the site configuration.
```
```suggestion
```

daxmc99 commented:
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.18 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

mrnugget commented:
@sourcegraph/campaigns @sqs I rebased this PR onto `main` so that it's ready to be merged, once we're comfortable merging it. I also updated the PR description, but here's what it does:

- It replaces all the feature flags with a single `campaigns.enabled` flag that enables/disables campaigns for _all users_. It's on by default.
- I left in the safety switch: creating `changesetSpecs`/`campaignSpecs` (which are the entry points to all other write-operations in campaigns) is only allowed for admins.

That means: by merging this campaigns are enabled by default but only admins can create campaigns. I think that's fine. 

@eseliger due to the large rebase across 1.5 months of commits I lost the changes in the frontend. Could you take a quick look?

marekweb commented:
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.19 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

mrnugget commented:
Rebased and force-pushed.

mrnugget commented on doc/user/campaigns/managing_access.md:
```diff
@@ -17,6 +15,8 @@ To see the campaign's proposed changes on a repository, a person *also* needs re
 
 Site admins have admin permissions on all campaigns.
 
+For now, users only have read access to campaigns. In the future, users will have admin permissions on their own campaigns too.
```
This directly refers to the distinction between the two permission levels a user can have for a given campaign: "Read" and "Admin" , see the list further above. I don't want to be imprecise by introducing another capability here.

mrnugget commented on enterprise/internal/campaigns/resolvers/resolver.go:
```diff
@@ -204,12 +219,13 @@ func (r *Resolver) ApplyCampaign(ctx context.Context, args *graphqlbackend.Apply
 	opts := ee.ApplyCampaignOpts{}
 
 	opts.CampaignSpecRandID, err = unmarshalCampaignSpecID(args.CampaignSpec)
-	if err != nil {
-		return nil, err
-	}
```
Good catch! I think that got sucked into a time/space fracture created by a `git rebase`.
